src/main/java/de/tum/www1/orion/connector/ide/build/OrionBuildConnector.kt

```diff
@@ -0,0 +1,124 @@
+import de.tum.www1.orion.dto.BuildLogFileErrorsDTO
+import de.tum.www1.orion.ui.browser.BrowserWebView
+import de.tum.www1.orion.util.cefRouter
+import org.cef.browser.CefBrowser
+import org.cef.browser.CefFrame
+import org.cef.callback.CefQueryCallback
+import org.cef.handler.CefLoadHandlerAdapter
+import org.cef.handler.CefMessageRouterHandlerAdapter
+import java.util.*
+import java.util.stream.Collectors
+
+class OrionBuildConnector(browserWebView: BrowserWebView) : OrionConnector(browserWebView), IOrionBuildConnector {
+    override fun buildAndTestLocally() {
+        ServiceManager.getService(project, OrionInstructorBuildUtil::class.java).runTestsLocally()
+    }
+
+    override fun onBuildStarted(exerciseInstructions: String) {
+        // Only listen to the first execution result
+        val testParser = ServiceManager.getService(project, OrionTestParser::class.java)
+        if (!testParser.isAttachedToProcess) {
+            val runManager = RunManager.getInstance(project)
+            val settings = runManager
+                    .createConfiguration("Build & Test on Artemis Server", OrionSubmitRunConfigurationType::class.java)
+            (settings.configuration as OrionRunConfiguration).triggeredInIDE = false
+            ExecutionUtil.runConfiguration(settings, DefaultRunExecutor.getRunExecutorInstance())
+            testParser.parseTestTreeFrom(exerciseInstructions)
+        }
+    }
+
+    override fun onBuildFinished() {
+        ServiceManager.getService(project, OrionTestParser::class.java).onTestingFinished()
+    }
+
+    override fun onBuildFailed(buildLogsJsonString: String) {
+        val mapType = object : TypeToken<Map<String?, List<BuildError?>?>?>() {}.type
+        val allErrors = JsonParser().parse(buildLogsJsonString).asJsonObject["errors"]
+        val errors = Gson().fromJson<Map<String, List<BuildError>>>(allErrors, mapType)
+        val buildErrors = errors.entries.stream()
+                .map { fileErrors: Map.Entry<String, List<BuildError>> -> BuildLogFileErrorsDTO(fileErrors.key, fileErrors.value) }
+                .collect(Collectors.toList())
+        val testParser = ServiceManager.getService(project, OrionTestParser::class.java)
+        testParser.onCompileError(buildErrors)
+    }
+
+    override fun onTestResult(success: Boolean, testName: String, message: String) {
+        ServiceManager.getService(project, OrionTestParser::class.java).onTestResult(success, testName, message)
+    }
+
+    override fun initializeHandlers() {
+        jsQuery.cefRouter.addHandler(object : CefMessageRouterHandlerAdapter() {
+            override fun onQuery(browser: CefBrowser?, frame: CefFrame?, queryId: Long, request: String?, persistent: Boolean, callback: CefQueryCallback?): Boolean {
+                val scanner = Scanner(request)
+                val methodName = scanner.nextLine()
+                val methodNameEnum = try {
+                    IOrionBuildConnector.FunctionName.valueOf(methodName)
```

Comment:
Values with filter may look better. From a performance standpoint it doesn't matter much since it's just few names. If I have time I will do a profiling of the plugin to see which functions are hot and actually need to be optimized.

Reply:
The performance won't improve that much, true. My biggest issue was with the forced throw of an exception, I'd still suggest to choose the `values` option here

Reply:
Resolved.